test/integration/eds_integration_test.cc

```diff
@@ -437,5 +438,486 @@ TEST_P(EdsIntegrationTest, StatsReadyFilter) {
+    EXPECT_EQ(2, hosts_added.size());
+    member_update_count++;
+  });
+  setEndpoints(2, 2, 0);
+  EXPECT_EQ(1, member_update_count);
+}
+
+TEST_P(EdsOverGrpcIntegrationTest, StatsReadyFilter) {
+  config_helper_.prependFilter("name: eds-ready-filter");
+  initializeTest(false);
+  // Initial state: no healthy endpoints
+  EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_healthy")->value());
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      lookupPort("http"), "GET", "/cluster1", "", downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("500", response->headers().getStatusValue());
+  EXPECT_EQ("EDS not ready", response->body());
+  cleanupUpstreamAndDownstream();
+  // 2/2 healthy endpoints.
+  setEndpoints(2, 2, 0);
+  EXPECT_EQ(2, test_server_->gauge("cluster.cluster_0.membership_healthy")->value());
+  response = IntegrationUtil::makeSingleRequest(lookupPort("http"), "GET", "/cluster1", "",
+                                                downstream_protocol_, version_, "foo.com");
+  ASSERT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+  EXPECT_EQ("EDS is ready", response->body());
+  cleanupUpstreamAndDownstream();
+}
+
+TEST_P(EdsOverGrpcIntegrationTest, ReuseMuxAndStreamForMultipleClusters) {
+  if (edsUpdateMode() == Grpc::EdsUpdateMode::Multiplexed) {
+    initializeTest(false, 3, false);
+    EXPECT_EQ(0, test_server_->gauge("cluster.cluster_0.membership_total")->value());
+    EXPECT_EQ(0, test_server_->gauge("cluster.cluster_1.membership_total")->value());
+    EXPECT_EQ(0, test_server_->gauge("cluster.cluster_2.membership_total")->value());
+    switch (clientType()) {
+    case Grpc::ClientType::EnvoyGrpc:
+      // As EDS uses HTTP2, number of streams created by Envoy for EDS cluster equals to number
+      // of requests.
+      EXPECT_EQ(1UL, test_server_->counter("cluster.eds_cluster.upstream_rq_total")->value());
+      break;
+    case Grpc::ClientType::GoogleGrpc:
+      // One EDS mux/stream is created and reused for all 3 clusters when initializing first EDS
+      // cluster (cluster_0). As a consequence, only one Google async grpc client and one
+      // corresponding set of client stats should be created.
+      EXPECT_EQ(1UL,
+                test_server_->counter("cluster.cluster_0.grpc.eds_cluster.streams_total")->value());
+      EXPECT_EQ(TestUtility::findCounter(test_server_->statStore(),
+                                         "cluster.cluster_1.grpc.eds_cluster.streams_total"),
+                nullptr);
+      EXPECT_EQ(TestUtility::findCounter(test_server_->statStore(),
+                                         "cluster.cluster_2.grpc.eds_cluster.streams_total"),
+                nullptr);
+      break;
+    default:
+      PANIC("reached unexpected code");
+    }
+  } else {
+    test_skipped = true;
+  }
+}
+
```

Comment:
I'm wondering how difficult will it be to add a test that changes modes between "envoy.reloadable_features.multiplex_eds" enabled to disabled and back?

Reply:
@adisuissa it will require writing lots of code. Maybe not tricky, but large code wise